drivers/espi/espi_it8xxx2.c

```diff
@@ -687,22 +696,130 @@ static int espi_it8xxx2_write_lpc_request(const struct device *dev,
+		(struct espi_queue1_regs *)config->base_espi_queue1;
+	struct espi_oob_msg_packet *oob_pckt =
+		(struct espi_oob_msg_packet *)pckt->buf;
+	uint16_t oob_msg_len = oob_pckt->len_bit_7_0 |
+				((oob_pckt->tag_len_bit_11_8 & 0xf) << 8);
+
+	if (!(slave_reg->CH_OOB_CAPCFG3 & IT8XXX2_ESPI_OOB_READY_MASK)) {
+		LOG_ERR("%s: OOB channel isn't ready", __func__);
+		return -EIO;
+	}
 
-	return -EIO;
+	if (slave_reg->ESUCTRL0 & IT8XXX2_ESPI_UPSTREAM_BUSY) {
+		LOG_ERR("%s: OOB upstream busy", __func__);
+		return -EIO;
+	}
+
+	if (oob_msg_len > ESPI_IT8XXX2_OOB_MAX_PAYLOAD_SIZE) {
+		LOG_ERR("%s: Out of OOB queue space", __func__);
+		return -EINVAL;
+	}
+
+	/* Set cycle type */
+	slave_reg->ESUCTRL1 = IT8XXX2_ESPI_CYCLE_TYPE_OOB;
+	/* Set tag and length[11:8] */
+	slave_reg->ESUCTRL2 = oob_pckt->tag_len_bit_11_8;
+	/* Set length [7:0] */
+	slave_reg->ESUCTRL3 = oob_pckt->len_bit_7_0;
+
+	/* Set data byte */
+	for (int i = 0; i < oob_msg_len; i++) {
+		queue1_reg->UPSTREAM_DATA[i] = oob_pckt->data_byte[i];
+	}
+
+	/* Set upstream enable */
+	slave_reg->ESUCTRL0 |= IT8XXX2_ESPI_UPSTREAM_ENABLE;
+	/* Set upstream go */
+	slave_reg->ESUCTRL0 |= IT8XXX2_ESPI_UPSTREAM_GO;
+
+	return 0;
 }
 
 static int espi_it8xxx2_receive_oob(const struct device *dev,
 				struct espi_oob_packet *pckt)
 {
-	/* TODO: implement me... */
-	ARG_UNUSED(dev);
+	const struct espi_it8xxx2_config *const config = DRV_CONFIG(dev);
+	struct espi_it8xxx2_data *const data = DRV_DATA(dev);
+	struct espi_slave_regs *const slave_reg =
+		(struct espi_slave_regs *)config->base_espi_slave;
+	struct espi_queue0_regs *const queue0_reg =
+		(struct espi_queue0_regs *)config->base_espi_queue0;
+	struct espi_oob_msg_packet *oob_pckt =
+		(struct espi_oob_msg_packet *)pckt->buf;
+	int ret;
+	uint8_t oob_len;
+
+	if (!(slave_reg->CH_OOB_CAPCFG3 & IT8XXX2_ESPI_OOB_READY_MASK)) {
+		LOG_ERR("%s: OOB channel isn't ready", __func__);
+		return -EIO;
+	}
+
+	/* Wait until receive OOB message or timeout */
+	ret = k_sem_take(&data->oob_upstream_go, K_MSEC(ESPI_OOB_TIMEOUT_MS));
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		return -ETIMEDOUT;
+	}
+
+	/* Get length */
+	oob_len = (slave_reg->ESOCTRL4 & IT8XXX2_ESPI_PUT_OOB_LEN_MASK);
+	/*
+	 * Buffer passed to driver isn't enough.
+	 * The first three bytes of buffer are cycle type, tag, and length.
+	 */
+	if ((oob_len + 3) > pckt->len) {
+		LOG_ERR("%s: Out of rx buf %d vs %d",
+			__func__, (oob_len + 3), pckt->len);
+		return -EINVAL;
+	}
+
+	oob_pckt->cycle_type = ESPI_OOB_CYCLE_TYPE;
+	oob_pckt->tag_len_bit_11_8 = ESPI_OOB_TAG;
+	oob_pckt->len_bit_7_0 = oob_len;
 
-	return -EIO;
+	/* Get data byte */
+	for (int i = 0; i < oob_pckt->len_bit_7_0; i++) {
+		oob_pckt->data_byte[i] = queue0_reg->PUT_OOB_DATA[i];
+	}
+
+	return 0;
+}
+
+static void espi_it8xxx2_oob_init(const struct device *dev)
+{
+	const struct espi_it8xxx2_config *const config = DRV_CONFIG(dev);
+	struct espi_it8xxx2_data *const data = DRV_DATA(dev);
```

Comment:
As a follow up PR, please eliminate the DRV_CONFIG() and DRV_DATA() macros in the entire file.

Reply:
Ack.